### PR TITLE
Determine framework automatically before ONNX export

### PR DESCRIPTION
**Pull request by rachthree**
# What does this PR do?

Determines whether to use `torch` or `tf2onnx` as the ONNX exporter automatically with the following priority:
1. User input via `framework` / `--framework`.
2. If local checkpoint is provided, use the same framework as the checkpoint.
3. Available framework in environment, with priority given to PyTorch.

Fixes issue https://github.com/huggingface/transformers/issues/18495 where PyTorch was still attempted for a local TF checkpoint even though it did not exist in the environment. This also avoids requiring users to use `--framework=tf` when using the ONNX export driver script.

Misc:
* Adds `tf` to pip install for `run_tests_onnxruntime` and  `run_tests_onnxruntime_all` in CI.

## Tests
* `python -m transformers.onnx` driver with and without `--framework` on local checkpoints and hub. Tested in containerized environments that had only PyTorch, only TensorFlow, or both.
  * Successful.
* Unit tests: ran `RUN_SLOW=true pytest tests/onnx`
  ~~* Overall, tests **passed w.r.t `main`** since they share the same failing tests:~~
<strike>

```
FAILED tests/onnx/test_onnx.py::OnnxExportTestCase::test_quantize_pytorch - TypeError: 'module' object is not callable
FAILED tests/onnx/test_onnx.py::OnnxExportTestCase::test_quantize_tf - TypeError: 'module' object is not callable
FAILED tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_048_data2vec_vision_image_segmentation - ValueError: Unrecognized configuration class <class 'transformers.models.data2vec.configuration_data2vec_vision.Data2VecVisionConfig'> for this kind of AutoModel: AutoModelForImageSegmentation.
FAILED tests/onnx/test_onnx_v2.py::OnnxExportTestCaseV2::test_pytorch_export_on_cuda_048_data2vec_vision_image_segmentation - ValueError: Unrecognized configuration class <class 'transformers.models.data2vec.configuration_data2vec_vision.Data2VecVisionConfig'> for this kind of AutoModel: AutoModelForImageSegmentation.
```

</strike>

~~* Wrote up https://github.com/huggingface/transformers/issues/18614 for the `TypeError: 'module' object is not callable` errors.~~ **Fixed by https://github.com/huggingface/transformers/pull/18336**
~~* As for the `AutoModel` error, https://github.com/huggingface/transformers/blob/main/src/transformers/models/auto/modeling_auto.py#L363 says not to add new models, so is this failure acceptable?~~ **Fixed by https://github.com/huggingface/transformers/pull/18587**

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@LysandreJik and others who may be interested :)

### CODE REVIEW

**Review by lewtun — COMMENTED** [review]
Thank you so much for greatly improving the framework selection in the ONNX exporter @rachthree (also, welcome as a first time contributor 🥳)!

Overall, the logic looks great to me and I'd really like to see a unit test of the `determine_framework` function. This would give us some confidence that any future changes on the framework selection side won't accidentally break the desired behaviour.

Regarding the failing unit tests, these will be fixed by:

* https://github.com/huggingface/transformers/pull/18587
* https://github.com/huggingface/transformers/pull/18336

so we can rebase your branch on `main` once they're approved / merged (should be soon)

File: src/transformers/onnx/features.py
Comment (rachthree) [reply src/transformers/onnx/features.py]: Thanks! Done in [b832988](https://github.com/huggingface/transformers/pull/18615/commits/b83298803227cf8a79c2e0eb5a4724cf52275513)

File: src/transformers/onnx/features.py
Comment (rachthree) [reply src/transformers/onnx/features.py]: Thanks! Done in [b832988](https://github.com/huggingface/transformers/pull/18615/commits/b83298803227cf8a79c2e0eb5a4724cf52275513)

File: src/transformers/onnx/features.py
Comment (rachthree) [reply src/transformers/onnx/features.py:632]: Thanks! It helped me figure out the behavior, so hope it's helpful for others!

File: src/transformers/onnx/features.py
Comment (rachthree) [reply src/transformers/onnx/features.py:559]: Added a unit test in [b832988](https://github.com/huggingface/transformers/pull/18615/commits/b83298803227cf8a79c2e0eb5a4724cf52275513), but I put it under `tests/onnx/test_onnx_v2::OnnxUtilsTestCaseV2`. I just noticed you specified `test_features.py`, but it does not exist yet. I can create it if you'd like, or should I leave it as is?

File: src/transformers/onnx/features.py
Comment (rachthree) [reply src/transformers/onnx/features.py]: Thanks for the catch!

**Comment by rachthree** [comment]
> Thank you so much for greatly improving the framework selection in the ONNX exporter @rachthree (also, welcome as a first time contributor 🥳)!
> 
> Overall, the logic looks great to me and I'd really like to see a unit test of the `determine_framework` function. This would give us some confidence that any future changes on the framework selection side won't accidentally break the desired behaviour.
> 
> Regarding the failing unit tests, these will be fixed by:
> 
> * #18587
> * #18336
> 
> so we can rebase your branch on `main` once they're approved / merged (should be soon)

Thank you for the review and welcoming me! I'm excited to contribute, especially since this is my first PR in the open source community :)  Glad to see the 2 PRs will fix those unit tests.

File: tests/onnx/test_onnx_v2.py
Comment (lewtun) [inline tests/onnx/test_onnx_v2.py:136]: I suggest promoting this to a standalone class like `DetermineFrameWorkTest` and then treat each case with a dedicated function like `test_framework_provided(self)` etc

File: tests/onnx/test_onnx_v2.py
Comment (lewtun) [inline tests/onnx/test_onnx_v2.py:55]: The `transformers` codebase follows the (standard?) convention of using context managers with `TemporaryDirectory()` to handle temporary files / folders in our test suite. Could you please refactor to match this approach?

File: src/transformers/onnx/features.py
Comment (lewtun) [reply src/transformers/onnx/features.py:559]: Thank you!! Yes, please create a new `test_features.py` file for this test (we usually map `transformers/path/to/module.py` with `tests/path/to/test_module.py`)

File: tests/onnx/test_onnx_v2.py
Comment (lewtun) [inline tests/onnx/test_onnx_v2.py:52]: Do the tests also work with `SMALL_MODEL_IDENTIFIER` from `testing_utils.py`? That checkpoint is preferred since it's maintained by the HF team and less likely to vanish unexpectedly :)

**Review by lewtun — COMMENTED** [review]
Thank you very much for adding an extensive test suite for this functionality @rachthree ! I think we just need to do a little bit of refactoring and this should be good to go 🔥 !

File: tests/onnx/test_onnx_v2.py
Comment (rachthree) [reply tests/onnx/test_onnx_v2.py]: Didn't realize that was an available variable to use! That works, done in [67416f2](https://github.com/huggingface/transformers/pull/18615/commits/67416f22094bb29b7bdcfe22854ee00ef4e1e112).

File: tests/onnx/test_onnx_v2.py
Comment (rachthree) [reply tests/onnx/test_onnx_v2.py]: I originally did this manually because I didn't want files to be written again between unit tests, but it has been refactored in [67416f2](https://github.com/huggingface/transformers/pull/18615/commits/67416f22094bb29b7bdcfe22854ee00ef4e1e112) to match the approach.

File: tests/onnx/test_onnx_v2.py
Comment (rachthree) [reply tests/onnx/test_onnx_v2.py]: Thanks for the suggestion! Done in [67416f2](https://github.com/huggingface/transformers/pull/18615/commits/67416f22094bb29b7bdcfe22854ee00ef4e1e112). I broke it down to the 3 main paths, hope that is sufficient.

File: src/transformers/onnx/features.py
Comment (rachthree) [reply src/transformers/onnx/features.py:564]: Thanks! Done in [67416f2](https://github.com/huggingface/transformers/pull/18615/commits/67416f22094bb29b7bdcfe22854ee00ef4e1e112).

In [8da5990](https://github.com/huggingface/transformers/pull/18615/commits/8da5990fda344f7eb754bc8a9a0f44fac56b00fd) and  I registered the tests in `utils/tests_fetcher.py` because of a failure I got in CI saying that the test would not be discovered. Is this the correct way to add them?

In [63198fd](https://github.com/huggingface/transformers/pull/18615/commits/63198fd2b7c018cb2b20580cf1ec8614079e078f) I added `tf` for for the pip install steps for `run_tests_onnxruntime` and `run_tests_onnxruntime_all` in `.circleci/config.yml` so that `TFAutoModel` can be used. Also added `-rA` flags so that the results would be more verbose. In the logs for `run_tests_onnxruntime` it can be seen that the new unit tests are tested.

File: .circleci/config.yml
Comment (lewtun) [inline .circleci/config.yml:891]: I'm not sure if this change will interfere with our internal Slack reporting. Gently pinging @ydshieh for his input 🙏

File: src/transformers/onnx/features.py
Comment (lewtun) [reply src/transformers/onnx/features.py:564]: Thank you for registering the test - this is indeed the way to include it :)

File: tests/onnx/test_features.py
Comment (lewtun) [inline tests/onnx/test_features.py:57]: Nit: to save a bit of code duplication + ease of maintenance, we could move these variables to `setUp` and then access them with `self.torch_str`

We also tend to prefer putting the type on variable names, so an alternative would be to use something like `framework_pt = "pt"` and `framework_tf = "tf"` instead of `torch_str` etc

File: tests/onnx/test_features.py
Comment (lewtun) [inline tests/onnx/test_features.py:108]: turbo nit:
```suggestion
        # Both not in environment -> raise error
```

**Review by lewtun — APPROVED** [review]
Thank you very much for iterating on this @rachthree! The test coverage looks solid and I've just left some final nits - apart from that it LGTM 🔥 

Gently pinging @patrickvonplaten or @LysandreJik for final approval

File: .circleci/config.yml
Comment (ydshieh) [reply .circleci/config.yml:891]: When using `--make-reports`, there is no more need to use `-rA`.

See [here](https://github.com/huggingface/transformers/blob/6faf283288ce3390281ad8c1d37ccb13f2d03990/src/transformers/testing_utils.py#L1283)

I believe `-rA` will display the very detailed results on the terminal, which we would like to avoid. Instead, the summary is saved to files when we use `--make-reports` (this is not a pytest flag, but our own custom flag).

see [pytest doc](https://docs.pytest.org/en/7.1.x/how-to/output.html#producing-a-detailed-summary-report)

Comment (rachthree) [reply .circleci/config.yml:891]: @ydshieh I added `-rA` because even though the test was included in `tests_fetcher.py`, `require_tf` actually skipped the test since originally TensorFlow was not installed for `run_tests_onnxruntime`. I wouldn't have known that I needed to add `tf` for `run_tests_onnxruntime` without using `-rA` and checking that the tests I expected to run did run. The artifacts did not say exactly which tests were run, but what would be collected. Does `--make-reports` generate other reports that are not in CircleCI?

File: tests/onnx/test_features.py
Comment (rachthree) [reply tests/onnx/test_features.py]: Good idea, thanks! Done in [6bd7477](https://github.com/huggingface/transformers/pull/18615/commits/6bd7477fb2ddb5932ac46b0f59a945bfe3290d12).

File: .circleci/config.yml
Comment (ydshieh) [reply .circleci/config.yml:891]: @rachthree In `ARTIFACTS` tab of this [run_tests_onnxruntime run page](https://app.circleci.com/pipelines/github/huggingface/transformers/46219/workflows/f5b23910-e483-41c6-b313-054d23a620f5/jobs/542677/artifacts), you could find `~/transformers/tests_output.txt`. Click it, at the end, you will see something like
```bash
...
...
SKIPPED [1] tests/onnx/test_onnx.py:86: test is slow
SKIPPED [1] tests/onnx/test_onnx.py:75: test is slow
SKIPPED [400] ../.pyenv/versions/3.7.12/lib/python3.7/unittest/case.py:93: test is slow
SKIPPED [2] tests/onnx/test_onnx_v2.py:370: test is slow
================ 45 passed, 444 skipped, 45 warnings in 46.63s =================
```

In previous runs where TF was not installed, we should be able to see those ONNX tests were skipped for TF.

File: tests/onnx/test_features.py
Comment (lewtun) [reply tests/onnx/test_features.py]: Thanks!

File: .circleci/config.yml
Comment (rachthree) [reply .circleci/config.yml]: @ydshieh I went back to the CI run for commit [8da5990](https://github.com/huggingface/transformers/pull/18615/commits/8da5990fda344f7eb754bc8a9a0f44fac56b00fd) which didn't have TF and `-rA`, and even though [~/transformers/tests_output.txt](https://output.circle-artifacts.com/output/job/7f7e32ea-de63-485a-ac3a-2acccef11190/artifacts/0/~/transformers/tests_output.txt) didn't show the breakdown, [~/transformers/reports/tests_onnx/summary_short.txt](https://output.circle-artifacts.com/output/job/7f7e32ea-de63-485a-ac3a-2acccef11190/artifacts/0/~/transformers/reports/tests_onnx/summary_short.txt) did. Somehow I missed this when going through the artifacts, my apologies! `-rA` has been removed in [d8f3804](https://github.com/huggingface/transformers/pull/18615/commits/d8f38043867488ed61ed9d62175159acbcc09d1b). Thanks :)

There are other runs in `.circleci/config.yaml` that still have `-rA` from before... will the flag be removed from those runs in the future?

Comment (ydshieh) [reply .circleci/config.yml]: @rachthree  Thank you for double checking in commit 8da5990.

 @stas00 @LysandreJik Is there any reason to keep (or not) the flag `-rA` when we already use `--make-reports`?

Comment (stas00) [reply .circleci/config.yml]: It has been a long time since I wrote that code, does it still work with `-rA`? If so, there is no reason, but please remember that `--make-reports` is a hack, so it's quite possible that that flag was needed.

Please try w/o it and if it still produces the same output, then we may choose to remove it.

It's also possible that it is there if we wanted both - the normal log and the `--make-reports` individual logs.

Comment (rachthree) [reply .circleci/config.yml]: @stas00 For `run_tests_onnxruntime`, it still works. `~/transformers/reports/tests_onnx/summary_short.txt` always has the breakdown, so `-rA` is not needed as long as people know which artifact to look for. However, the output changes for the terminal output  in CircleCI and `~/transformers/tests_outputs.txt` as the short test summary info doesn't show without it. 

As for the rest of the tests in the CircleCI config, I think removing `-rA` would be outside the scope of this PR, but I can try it here or another branch if you'd like. Not having consistency confused me when I looked at the logs of the other tests.

Comment (ydshieh) [reply .circleci/config.yml]: @rachthree I think we can merge this PR first, with `-rA` if you prefer. As other tests also have this flag, I won't worry about having it.

We can check if there is a need to remove this flag in a separate PR, but this is not a priority.

File: src/transformers/onnx/features.py
Comment (patrickvonplaten) [inline src/transformers/onnx/features.py:613]: That looks like a breaking change to me -> why not leave `"pt"`?

**Review by patrickvonplaten — APPROVED** [review]
Cool, looks good to me. PyTorch has priority in the hierarchy when detecting the framework, so that's backwards compatible

File: .circleci/config.yml
Comment (stas00) [reply .circleci/config.yml]: Please do what's the best. I was just mentioning that if I remembered correctly some of those flags were still needed for `--make-reports` to work as intended.

Comment (rachthree) [reply .circleci/config.yml]: PR's been merged without the `-rA` flag, maintaining previous logging behavior. Thank you all for your review 😊

**Pull request merged**
